src/main/java/org/springframework/data/jpa/repository/query/QueryUtils.java

```diff
@@ -90,7 +90,7 @@
 	// Cc Control
 	// Cf Format
 	// P Punctuation
-	private static final String IDENTIFIER = "[._[\\P{Z}&&\\P{Cc}&&\\P{Cf}&&\\P{P}]]+";
```

Comment:
This adds a the `|` to the list of separators for an identifiers. 
But the underlying problem lies deeper: the Unicode punctuation (`\p{P}`)category doesn't contain many of the values in the Posix punctuation category (`\{Punct}`). For example `<` and `>` aren't in that list either.

I quickly tried `\{Punct}` but breaks some other test.

In any case we need a solution addressing the general problem.

Reply:
Thanks for the feedback @schauder , I have refactored the implementation a bit to address the issue at large, wondering what do you think, all tests are passing. Thank you!

Reply:
While this fixes the problem at hand while not breaking any tests it just makes it obvious that something is really wrong here: an Identifier shouldn't contain a `#` and also probably not a `$`.

It seems `IDENTIFIER` isn't used for matching identifier, but something like identifier-or-spel 
Therefore it

a) should be named as such
b) should be constructed as such, i.e. out of two regex that matches identifiers and spels respectively.
c) all the usages should be verified that they actually should use the more general regex vs the one that actually matches identifiers.

Reply:
So the `$` seems to be a valid symbol within identifier name, there is even a dedicated test for it:
```
@Test // DATAJPA-1200
public void testHasNamedParameter() {
    ...
    checkHasNamedParameter(":$something", true, "dollar");
    ...
}
```

The `#` indeed is the SpelEx delimiter, in this context it turned out to be correct but redundant: it is explicitly included  within `NAMED_PARAMETER` pattern. So I removed it from the more strict `IDENTIFIER` match.